server/modlog.ts

```diff
@@ -36,284 +31,241 @@ const PUNISHMENTS = [
 	'NOTE', 'MAFIAHOSTBAN', 'MAFIAUNHOSTBAN', 'GIVEAWAYBAN', 'GIVEAWAYUNBAN',
 	'TOUR BAN', 'TOUR UNBAN', 'UNNAMELOCK',
 ];
-const PUNISHMENTS_REGEX_STRING = `\\b(${PUNISHMENTS.join('|')}):.*`;
-
-const execFile = util.promisify(child_process.execFile);
 
 export type ModlogID = RoomID | 'global';
 
 interface ModlogResults {
-	results: string[];
+	results: ModlogEntry[];
 	duration?: number;
 }
 
 interface ModlogQuery {
-	rooms: ModlogID[];
-	search: string;
-	isExact: boolean;
-	maxLines: number;
-	onlyPunishments: boolean | string;
+	query: string;
+	args: any[];
 }
 
-class SortedLimitedLengthList {
-	maxSize: number;
-	list: string[];
-
-	constructor(maxSize: number) {
-		this.maxSize = maxSize;
-		this.list = [];
-	}
+export interface ModlogSearch {
+	note?: {searches: string[], isExact?: boolean};
+	user?: string;
+	ip?: string;
+	action?: string;
+	actionTaker?: string;
+}
 
-	getListClone() {
-		return this.list.slice();
-	}
+export interface ModlogEntry {
+	action: string;
+	roomID?: string;
+	visualRoomID?: string;
+	userid?: ID;
+	autoconfirmedID?: ID;
+	alts?: ID[];
+	ip?: string;
+	isGlobal?: boolean;
+	loggedBy?: ID;
+	note?: string;
+	/** Milliseconds since the epoch */
+	time?: number;
+}
 
-	insert(element: string) {
-		let insertedAt = -1;
-		for (let i = this.list.length - 1; i >= 0; i--) {
-			if (element.localeCompare(this.list[i]) < 0) {
-				insertedAt = i + 1;
-				if (i === this.list.length - 1) {
-					this.list.push(element);
-					break;
-				}
-				this.list.splice(i + 1, 0, element);
-				break;
-			}
-		}
-		if (insertedAt < 0) this.list.splice(0, 0, element);
-		if (this.list.length > this.maxSize) {
-			this.list.pop();
-		}
+export class Modlog {
+	readonly database: Database.Database;
+	readonly usePM: boolean;
+
+	constructor(path: string, noProcessManager = false) {
+		this.database = new Database(path);
+		this.database.exec(FS(`${__dirname}/../databases/schemas/modlog-schema.sql`).readIfExistsSync()); // Set up tables, etc.
+		this.database.function('regex', {deterministic: true}, (regexString, toMatch) => {
+			return Number(RegExp(regexString).test(toMatch));
+		});
+		this.usePM = !noProcessManager;
 	}
-}
 
-export function checkRipgrepAvailability() {
-	if (Config.ripgrepmodlog === undefined) {
-		Config.ripgrepmodlog = (async () => {
+	async runSQL(query: ModlogQuery, forceNoPM = false): Promise<Database.RunResult | any[]> {
+		if (!forceNoPM && this.usePM) {
+			return PM.query(query);
+		} else {
+			const statement = this.database.prepare(query.query).bind(...query.args);
```

Comment:
Slow to prepare a statement every time you query. Queries should be prepared once and reused.

Reply:
I can prepare most of the queries on initialization, but since search queries are dynamically generated based on which fields are searched for, I think that preparing all possible combinations would be rather cumbersome. Maybe have a cache in the form of a string:statement map?